src/sparkdantic/model.py

```diff
@@ -222,110 +178,115 @@ def _get_enum_mixin_type(t: EnumType) -> MixinType:
         raise TypeError(f'Enum {t} is not supported. Only int and str mixins are supported.')
 
 
-def _type_to_spark(t: Type, metadata, safe_casting: bool = False) -> Tuple[DataType, bool]:
-    """Converts a given Python type to a corresponding PySpark data type.
+def _from_python_type(
+    type_: Type,
+    metadata: list[Any],
+    safe_casting: bool = False,
+) -> spark_types.DataType:
+    """Converts a Python type to a corresponding PySpark data type.
 
     Args:
-        t (Type): The type to convert to a PySpark data type.
+        type_ (Type): The python type to convert to a PySpark data type.
+        metadata (list): The metadata for the field.
+        safe_casting (bool): Indicates whether to use safe casting for integer types.
 
     Returns:
         DataType: The corresponding PySpark data type.
     """
-    nullable, t = _is_nullable(t)
-    meta = None if len(metadata) < 1 else deepcopy(metadata).pop()
-    args = get_args(t)
-    origin = get_origin(t)
+    py_type = _get_union_type_arg(type_)
+
+    if _is_base_model(py_type):
+        return create_spark_schema(py_type, safe_casting)
+
+    args = get_args(py_type)
+    origin = get_origin(py_type) or py_type
```

Comment:
Upon review, this is misleading. It's due to `origin` being `None` for incomplete type annotations on `list` and `dict` (and `Literal`). So it would be better to do something like:

```python
if origin is None and py_type in (list, dict):
    raise ...
```

Will fix this up